storm-core/src/jvm/backtype/storm/blobstore/KeySequenceNumber.java

```diff
@@ -0,0 +1,227 @@
+            }
+
+            // Update scenario 2 and 3 explain the code logic written here
+            // especially when nimbus crashes and comes up after and before update
+            // respectively.
+            int currentSeqNumber = getMaxSequenceNumber(zkClient);
+            if (!checkIfStateContainsCurrentNimbusHost(stateInfoList, nimbusInfo) && !nimbusInfo.isLeader()) {
+                if (sequenceNumbers.last() < currentSeqNumber) {
+                    return currentSeqNumber;
+                } else {
+                    return 0;
+                }
+            }
+
+            // It covers scenario's expalined in scenario 3 when nimbus-1 holding the latest
+            // update goes down before it is downloaded by nimbus-2. Nimbus-2 gets elected as a leader
+            // after which nimbus-1 comes back up and a read or update is performed.
+            if (!checkIfStateContainsCurrentNimbusHost(stateInfoList, nimbusInfo) && nimbusInfo.isLeader()) {
+                incrementMaxSequenceNumber(zkClient, currentSeqNumber);
+                return currentSeqNumber + 1;
+            }
+
+            // This code logic covers the update scenarios in 2 when the nimbus-1 goes down
+            // before syncing the blob to nimbus-2 and an update happens.
+            // If seq-num for nimbus-2 is 2 and max-seq-number is 3 then next sequence number is 4
+            // (max-seq-number + 1).
+            // Other scenario it covers is when max-seq-number and nimbus seq number are equal.
+            if (sequenceNumbers.size() == 1) {
+                if (sequenceNumbers.first() < currentSeqNumber) {
+                    incrementMaxSequenceNumber(zkClient, currentSeqNumber);
+                    return currentSeqNumber + 1;
+                } else {
+                    incrementMaxSequenceNumber(zkClient, currentSeqNumber);
+                    return sequenceNumbers.first() + 1;
+                }
+            }
+        } catch(Exception e) {
+            LOG.error("Exception {}", e);
+        } finally {
+            if (zkClient != null) {
+                zkClient.close();
+            }
+        }
+        // Normal create update sync scenario returns the greatest sequence number in the set
+        return sequenceNumbers.last();
+    }
+
+    private boolean checkIfStateContainsCurrentNimbusHost(List<String> stateInfoList, NimbusInfo nimbusInfo) {
+        boolean containsNimbusHost = false;
+        for(String stateInfo:stateInfoList) {
+            if(stateInfo.contains(nimbusInfo.getHost())) {
+                containsNimbusHost = true;
+                break;
+            }
+        }
+        return containsNimbusHost;
+    }
+
+    private void incrementMaxSequenceNumber(CuratorFramework zkClient, int count) throws Exception {
+        zkClient.setData().forPath(BLOBSTORE_MAX_KEY_SEQUENCE_SUBTREE + "/" + key,
+                ByteBuffer.allocate(INT_CAPACITY).putInt(count++).array());
```

Comment:
`count++` this is a post-increment operator.  This will write `count`'s value that was given by the caller, then increment the value, then throw it away.  It looks like we want `count+1` here instead.